internal/controller/pipelinerollout/pipelinerollout_controller.go

```diff
@@ -1118,10 +1118,22 @@ func (r *PipelineRolloutReconciler) Recycle(ctx context.Context,
 
 		case common.LabelValueProgressiveSuccess:
 			pause = true
-			requireDrain = true // TODO: make configurable (https://github.com/numaproj/numaplane/issues/512)
+			// TODO: make configurable (https://github.com/numaproj/numaplane/issues/512)
+			requireDrain = true
 
 		case common.LabelValueProgressiveFailure:
+			// this is the case of a previously failed "upgrading" pipeline which was replaced with a new one
+		}
+	}
 
+	if requireDrain {
+		modified, err := r.ensurePipelineIsDrainable(ctx, pipeline)
+		if err != nil {
+			return false, err
+		}
+		if modified {
+			// come back later
```

Comment:
Should we add "TODO" here to easily find later?

Reply:
sorry, that didn't mean that we need to come back later to look at it, it just meant that we are requeuing here - let me know if I should improve the comment

Reply:
okay, It's sounds confusing, maybe we can mention "requeuing" also in the comment

Reply:
will do